src/util-rust.c

```diff
@@ -0,0 +1,218 @@
+/* Copyright (C) 2007-2017 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ *
+ * \file
+ *
+ * \author Pierre Chifflier <chifflier@wzdftpd.net>
+ */
+
+#include "suricata-common.h"
+#include "debug.h"
+#include "detect.h"
+#include "pkt-var.h"
+#include "conf.h"
+
+#include "util-print.h"
+#include "util-debug.h"
+
+#include "app-layer-detect-proto.h"
+#include "app-layer-parser.h"
+
+#include "util-rust.h"
+
+#ifndef HAVE_RUST
+
+/* Rust support not available */
+void RustInit(void)
+{
+}
+
+#else /* HAVE_RUST */
+
+static int RustRegisterParser(const struct rust_parser * parser);
+
+static struct rust_config _rcfg = {
+    .magic = 0x72757374,
+};
+
+// Helper function to discover the ALPROTO identifier from the Rust code
+//
+// Later, this cpde will be replaced by the dynamic registration of ALPROTO
+AppProto crust_register_alproto(const char *proto)
+{
+    if (strcmp("rust-ntp",proto)==0)
+        return ALPROTO_NTP;
+    else if (strcmp("rust-tls",proto)==0)
+        return ALPROTO_TLS;
+    else {
+        SCLogError(SC_ERR_INITIALIZATION, "Unknown Rust protocol");
+        return -1;
+    }
+
+}
+
+
+static int RustParseToServer(Flow *f, void *state,
```

Comment:
I like how this would make the C-side of things much leaner. I like to see what we learn from the first couple of 'hard coded' protocols to figure out how to make this flexible enough.